src/category_theory/lifting_properties.lean

```diff
@@ -0,0 +1,208 @@
+property with respect to arrows in a given diagram.
+
+## Main results
+- `has_lifting_property`: the definition of the lifting property
+- `iso_has_right_lifting_property`: any isomorphism satisfies the right lifting property (rlp)
+- `id_has_right_lifting_property`: any identity has the rlp
+- `right_lifting_property_initial_iff`: spells out the rlp against a map whose source is an 
+initial object
+- `right_lifting_subcat`: given a functor F : D ⥤ arrow C, we construct the subcategory of those 
+morphisms p in C that satisfy the rlp w.r.t. F(i), for any object i in D.
+
+## Tags
+lifting property
+-/
+
+namespace category_theory
+
+universes v u
+variables {C : Type u} [category.{v} C]
+variables {D : Type u} [category.{v} D]
+
+variables {X Y Z : C}
+
+/- The left lifting property of a morphism i vs. a morphism p. -/
+def has_lifting_property (i p : arrow C) : Prop := ∀ (sq : i ⟶ p), arrow.has_lift sq
+
+/- Any isomorphism has the right lifting property with respect to any map. 
+A    → X 
+↓i    ↓p≅
+B    → Y
+-/
+lemma iso_has_right_lifting_property (i : arrow C) (p : X ≅ Y) : 
+has_lifting_property i (arrow.mk p.hom) :=
+begin
+  intro sq,
+  fconstructor,
+  fconstructor,
+  fconstructor,
+  { use sq.right ≫ p.inv }, -- the lift is obtained by p⁻¹ ∘ (B → Y)
+  { rw [←category.assoc, (arrow.w sq).symm],
+    simp, },
+  { simp, }
+end
+
+/- Any identity has the right lifting property with respect to any map. -/
+lemma id_has_right_lifting_property (i : arrow C) : has_lifting_property i (arrow.mk (𝟙 X)) :=
+begin
+  haveI := is_iso.id X,
+  exact iso_has_right_lifting_property i (category_theory.as_iso (𝟙 X)),
+end
+
+/- An equivalent characterization for right lifting against a map i whose source is initial. 
+∅ → X
+↓   ↓
+B → Y has a lifting iff there is a map B → X making the right part commute.
+-/
+lemma right_lifting_property_initial_iff (i p : arrow C) 
+(h : category_theory.limits.is_initial i.left) :
+has_lifting_property i p ↔ ∀ {e : i.right ⟶ p.right}, ∃ l : i.right ⟶ p.left, l ≫ p.hom = e :=
+begin
+  split,
+  { intros hlift bottom,
+    have comm : (category_theory.limits.is_initial.to h p.left) ≫ p.hom = i.hom ≫ bottom := 
+    begin apply category_theory.limits.is_initial.hom_ext h, end,
+    haveI := hlift (arrow.hom_mk comm),
+    use (arrow.has_lift.struct (arrow.hom_mk comm)).lift,
+    rw (arrow.has_lift.struct (arrow.hom_mk comm)).fac_right,
+    refl },
+  { intros h1 sq,
+    cases h1 with e he,
+    fconstructor,
+    fconstructor,
+    fconstructor,
+    { exact e },
+    { apply category_theory.limits.is_initial.hom_ext, simp, exact h },
+    { exact he } }
+end
+
+/- A helper construction: given a square between i and f ≫ g, produce a square between i and g, 
+whose top leg uses f:
+A  → X  
+     ↓f
+↓i   Y             --> A → Y 
+     ↓g                ↓i  ↓g
+B  → Z                 B → Z 
+ -/
+def square_to_second {i : arrow C} {f : X ⟶ Y} {g : Y ⟶ Z} (sq : i ⟶ arrow.mk (f ≫ g)) :
```

Comment:
```suggestion
@[simps] def square_to_second {i : arrow C} {f : X ⟶ Y} {g : Y ⟶ Z} (sq : i ⟶ arrow.mk (f ≫ g)) :
```
This will help in the proof of `has_right_lifting_property_comp`.

Reply:
@TwoFX I have now added this, but I am not sure how exactly it helps there. Currently, in the proof of `has_right_lifting_property_comp`, I have this:

```
have : sq0.right = sq1.right := rfl,
    rw this,
```

I suspect the purpose of the @simps label is that I might just as well omit this, but this doesn't compile. How would I make this work?

Reply:
The idea of the `simps` label is to let the simplifier "unfold" `square_to_snd`. Unfortunately, the simplifier won't unfold names created by the `let` tactic, so you inadvertently negated the benefit I had in mind. Here's my suggestion: if you add the lemma
```lean
@[simp, reassoc]
lemma lift.fac_right_of_to_mk {X Y : T} {f : arrow T} {g : X ⟶ Y} (sq : f ⟶ mk g)
  [has_lift sq] : lift sq ≫ g = sq.right :=
by simp_rw [←mk_hom g, lift.fac_right]
```
after `lift.fac_right` in `arrow.lean`, then the proof of `has_right_lifting_property_comp` reduces to
```lean
  intro sq0, -- a square between i and f ≫ g

  -- lift i vs. g
  haveI := hg (arrow.square_to_snd sq0),

  -- construct a square i ⟶ f
  let sq2 : i ⟶ (arrow.mk f) :=
  { left := sq0.left,
    right := arrow.lift (arrow.square_to_snd sq0) },

  -- construct a lift i vs. f
  haveI := hf sq2,

  -- show that this lift is a lift of i vs. g ∘ f
  refine ⟨⟨{lift := (arrow.lift sq2 : _ ⟶ _), fac_right' := _}⟩⟩,
  { simp }
```
(and the two final lines can then subsequently be condensed to `exact ⟨⟨⟨(arrow.lift sq2 : _ ⟶ _), by simp⟩⟩⟩`).

Reply:
I just opened #6953, which contains the lemma mentioned in my previous comment. You can "import" the new lemma from my PR into your PR by saying `git fetch` followed by `git merge origin/arrow_mk` and editing the top comment in this PR to mention the new dependency.

Reply:
Thanks a lot for your help! This is now implemented.